pkg/util/provider/machinecontroller/machine_util.go

```diff
@@ -708,31 +718,78 @@ func (c *controller) reconcileMachineHealth(ctx context.Context, machine *v1alph
 					machine.Name,
 					timeOutDuration,
 				)
+				// Log the error message for machine failure
+				klog.Error(description)
+
+				clone.Status.LastOperation = v1alpha1.LastOperation{
+					Description:    description,
+					State:          v1alpha1.MachineStateFailed,
+					Type:           machine.Status.LastOperation.Type,
+					LastUpdateTime: metav1.Now(),
+				}
+				clone.Status.CurrentStatus = v1alpha1.CurrentStatus{
+					Phase: v1alpha1.MachineFailed,
+					//TimeoutActive:  false,
+					LastUpdateTime: metav1.Now(),
+				}
+				objectRequiresUpdate = true
 			} else {
-				// Timeour occurred due to machine being unhealthy for too long
+
+				// Timeout occurred due to machine being unhealthy for too long
 				description = fmt.Sprintf(
 					"Machine %s is not healthy since %s minutes. Changing status to failed. Node Conditions: %+v",
 					machine.Name,
 					timeOutDuration,
 					machine.Status.Conditions,
 				)
-			}
 
-			// Log the error message for machine failure
-			klog.Error(description)
+				machineSetName, machineDeployName := getMachineOwnerNames(machine)
+				//getting lock for machineDeployment this machine belongs to, creating lock if absent
+				tempCh := make(chan struct{}, 1)
+				val, loaded := c.safetyOptions.HealthChanMap.LoadOrStore(machineDeployName, tempCh)
+				if loaded {
+					close(tempCh)
+				} else {
+					klog.Infof("Lock for machineDeploy=%s created", machineDeployName)
+				}
+				ch := val.(chan struct{})
+
+				klog.Errorf("Number of goroutines now %d\n", runtime.NumGoroutine())
+				lockAcquireCtx, cancelFn := context.WithTimeout(ctx, lockAcquireTimeout)
+				defer cancelFn()
+
+				select {
+				case ch <- struct{}{}:
+					klog.Infof("Acquired lock ,machineName=%q , locks left to acquire %d", machine.Name, cap(ch)-len(ch))
+					markable, err := c.canMarkMachineFailed(machineSetName, machineDeployName, machine.Namespace, maxReplacements)
+					if err != nil {
+						klog.Errorf("Couldn't check if machine can be marked as Failed. Error: %q", err)
+						<-ch
+						return machineutils.ShortRetry, err
+					}
+					if markable {
+						retryPeriod, updated, err := c.updateMachineToFailedState(ctx, description, machine, clone)
+						//wait for cache sync
+						if updated && !c.waitForFailedMachineCacheUpdate(machine, cacheUpdateTimeout) {
+							//waiting 1 sec since nothing else can be done if cache update is failing
+							klog.Infof("cache sync returned false, waiting 1 sec , machineName=%q", machine.Name)
+							<-time.After(1 * time.Second)
```

Comment:
```suggestion
							time.Sleep(1 * time.Second)
```
for better readability, as `<-time.After` only has an advantage if it used as a select case.

Reply:
In `waitForFailedMachineCacheUpdate()` I explicitly wait for the cache update as the reflector cache update is not that reliable. If it doesn't get updated in `waitForFailedMachineCacheUpdate()` (either due to timeout or any error) , then I wait for this `1sec` and thus depending only on the reflector update now.

Reply:
addressed in 2aa4416f08be317fc00c9b12355fe15ea986b5a8

Reply:
increased the wait time to 10sec to deal with it currently.